### PR TITLE
AWS SSM Connection Plugin Integration Tests

### PR DESCRIPTION
**Pull request by gau1991**
##### SUMMARY
This Pull Request covers the integration test for for AWS SSM Connection Plugin, introduced in the PR https://github.com/ansible/ansible/pull/49652

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
connection/aws_ssm

##### ADDITIONAL INFORMATION

This PR adds a integration test for AWS SSM Connection Plugin. 

For execution following command can be used:
```bash
test/runner/ansible-test integration -v connection_aws_ssm --allow-unsupported
```
The integration tests are still marked as `unsupported` as they are failing with following issue:
```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: UnicodeEncodeError: 'ascii' codec can't encode characters in position 5-6: ordinal not in range(128)
fatal: [test]: FAILED! => {"msg": "Unexpected failure during module execution.", "stdout": ""}
```
This need to be fixed from the AWS SSM Connection Plugin side.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -90,6 +90,7 @@ packaging/release/ansible_release
 /test/integration/inventory.remote
 /test/integration/inventory.networking
 /test/integration/inventory.winrm
+/test/integration/inventory.aws_ssm
```
Comment (mattclay) [inline .gitignore]: Revert this change.

File: test/integration/inventory.aws_ssm.template [hunk 2]
```diff
@@ -0,0 +1,11 @@
+[aws_ssm]
```
Comment (mattclay) [inline test/integration/inventory.aws_ssm.template]: Move this to `test/integration/targets/connection_aws_ssm/inventory`.

File: test/runner/lib/cli.py [hunk 3]
```diff
@@ -28,6 +28,7 @@
 
 from lib.executor import (
     command_posix_integration,
+    command_aws_ssm_integration,
```
Comment (mattclay) [inline test/runner/lib/cli.py]: Revert this change, and all other changes to `test/runner/lib/`.

**Review by mattclay — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 1 above)
Comment (gau1991) [reply .gitignore]: Hi @mattclay , 

I think we will need this file, as AWS SSM connection plugin requires the custom inventory file, which has informatiom about instance id, region and s3 bucket. 

Can you please advise, if we remove this from here, what will be better way of adding that?

Comment (mattclay) [reply .gitignore]: Tests for AWS should include the line `cloud/aws` in the `aliases` file for that test target. That will provide the necessary credentials for connecting to AWS. I recommend taking a look at the [inventory_aws_ec2](https://github.com/ansible/ansible/tree/devel/test/integration/targets/inventory_aws_ec2/) integration test to see how this works.

Comment (gau1991) [reply .gitignore]: Thanks for reply,

I think I am confused here. Although this is AWS SSM connection plugin, it is pretty much alternative SSH.

Also, this will use custom inventory file, which has instance id. So we need to create those instances before executing tests cases.  As each and every time instance id will change, using template for inventory `/test/integration/inventory.aws_ssm.template` is good approch instead of adding instance id into `/test/integration/targets/connection_aws_ssm/inventory` file

Let me know your thoughts on this.

Comment (mattclay) [reply .gitignore]: The test should create the instance it needs, run the AWS SSM connection tests, and then remove the instance. You can use a template for the inventory file as part of this process, but any files need by the test should reside under `test/integration/targets/connection_aws_ssm/`.

Comment (gau1991) [reply .gitignore]: Got it @mattclay, thanks for clarificaiton.  I do have last question before updating files. 

Lets say, I want to execute ping module integration test using AWS SSM Connection plugin. Right now it allows me using command:

```
test/runner/ansible-test aws-ssm-integration -v ping
```

How we can achieve this similar without doing changes in `test/runner/lib/` code?

Comment (mattclay) [reply .gitignore]: The standard connection plugin integration tests should cover everything needed for basic testing of the connection plugin, so running other integration tests using the connection plugin shouldn't be needed.

That being said, there are some upcoming changes to ansible-test which will allow running integration tests using connection plugins other than `local`, at which point you should be able to run arbitrary tests like `ping` using AWS SSM.

Comment (gau1991) [reply .gitignore]: Got it. Thanks for reply @mattclay . I will revert all the changes mentioned.

File: test/runner/lib/cli.py (hunk 3 above)
Comment (gau1991) [reply test/runner/lib/cli.py]: Done.

File: test/integration/inventory.aws_ssm.template (hunk 2 above)
Comment (gau1991) [reply test/integration/inventory.aws_ssm.template]: Done

File: .gitignore (hunk 1 above)
Comment (gau1991) [reply .gitignore]: @mattclay Reverted all mentioned changes.

File: test/integration/targets/connection_aws_ssm/aliases [hunk 4]
```diff
@@ -0,0 +1,2 @@
+unsupported
```
Comment (mattclay) [inline test/integration/targets/connection_aws_ssm/aliases:1]: The test can be supported if you update the `runme.sh` to run a playbook that creates an EC2 instance with the SSM Agent installed and then write out an inventory file for the rest of the test to use.

Comment (gau1991) [reply test/integration/targets/connection_aws_ssm/aliases:1]: Ok got it, will add support for that. 

Currently I added `unsupported` flag as the standerd connection test are failing for SSM Connection plugin because of unicode characters. We are in process to fix the plugin so that it will support all the unicode characters.

Comment (gau1991) [reply test/integration/targets/connection_aws_ssm/aliases:1]: Hi @mattclay, 
We updated `runme.sh` which will create Linux and Windows EC2 Instaces with SSM Agent Installed, and create inventory file as well for integration test cases execution.

Can you please review the code?

Comment (mattclay) [reply test/integration/targets/connection_aws_ssm/aliases:1]: Have the unicode issues been fixed? If so, remove the `unsupported` alias so the tests will run in CI.

If not, is there an issue or PR for fixing the issues that you can link to?

Comment (mattclay) [reply test/integration/targets/connection_aws_ssm/aliases:1]: You'll also need to add `shippable/aws/group1` to the aliases when removing `unsupported`.

**Comment by mattclay** [comment]
These tests should be in the same PR as the connection plugin since it hasn't been merged yet.

File: test/integration/targets/connection_aws_ssm/aliases (hunk 4 above)
Comment (gau1991) [reply test/integration/targets/connection_aws_ssm/aliases:1]: Sure. @mattclay . We already did this change in the PR https://github.com/ansible/ansible/pull/49652 . So closing this PR, as all the integration tests releated changes are cherry-picked.
